app/models/operatingsystem.rb

```diff
@@ -124,7 +124,8 @@ def repos(host)
   end
 
   def medium_uri(host, url = nil)
-    url ||= host.medium.path
+    url ||= host.medium.path unless host.medium.nil?
```

Comment:
Here and other places -- I would suggest using medium instead of host as the first parameter. Less coupling, easier to understand, etc.

Reply:
Two things:
1. It would be problematic for katello - they are not using our medium model when the host is consuming managed content. The URI in their case is determined by their own set of fields (`content_source`, `kt_environment` e.t.c )
2. I tried not to change method signatures, since I don't know who is using those methods outside foreman's core.

Reply:
what about not passing the medium, but just the `medium_url`

Reply:
This is the method that actually generates the URL, I don't think I can pass it in...

Reply:
oh, right, I was referring to the other methods, that we are adding the additional `host` arg now.

Reply:
That method (`pxe_prefix`) now calls implicitly to `medium_uri` for generating the digest

Reply:
That's why I would like rather to see some `PxeFilesService` object, that would encapsulate the paths and urls generation, and there could ve even multiple implementations (Foreman vs. Katello) that could influence where the paths and urls are. This would clean also the Operatingsystem class. So instead of `host.operatingsystem.{kernel(host),medium_url(architecture, host$,...}`, we could have `host.pxe_files.{kernel_path,medium_url,...}` (notice we would not need to pass the additional args to methods, as the `PxeFilesService` would choose which host/hostgroup attributes are relevant for it.

Reply:
Indeed we can do it, but it will require refactoring for both foreman and katello. Do we really want to get into this refactoring right now?

Reply:
If we don't do it now, then when? It's the ideal time from release perspective to do it now. Lets keep the current version around (in case it would turn out that the refactoring is no go), but let shoot for something better.

Reply:
OK I looked into it a bit more. This is what I have right now: there are two things mixed in OS class right now: URI providers (medium_url for example) and boot filenames (like `images/pxeboot/vmlinuz`). The former has nothing to do with OS, so it could be extended into a provider. The latter is OS-dependent so it should stay at the OS class.
So this feature actually falls into the latter category and should stay in OS, I would suggest creating a separate issue to move the former into proper providers pattern.

Reply:
Aren't they related in a way, that the end path (such as vmlinuz) is influenced by the `medium_url`? While I'm ok the base for the boot filenames to be provided by the OS, if we move the final construction of the path to the provider class, it would IMO clean up the code quite a lot, savings us passing some unnecessary args to the os methods.

Reply:
Before this PR they were completely unrelated - each OS "knows" what limitations exist on file names, and provides the paths accordingly. Provider is completely unaware where the file will be served, so I don't see a reason to push this responsibility to the provider.

Just to illustrate the issue, I have a host with CentOS set with installation medium set to the one supplied with Foreman.
`medium_url` is set to `http://mirror.centos.org/centos/7/os/x86_64` and `boot_files_uri` returns `[#<URI::HTTP http://mirror.centos.org/centos/7/os/x86_64/images/pxeboot/vmlinuz>,
 #<URI::HTTP http://mirror.centos.org/centos/7/os/x86_64/images/pxeboot/initrd.img>]`
The only method that actually connects the path as should be set on our tftp server with those paths is `pxe_files`:
`[{:"boot/Centos-7-x86_64-4b42362509125324ee277c89a503e04323ce5a70"=>
   "http://mirror.centos.org/centos/7/os/x86_64/images/pxeboot/vmlinuz"},
 {:"boot/Centos-7-x86_64-4b42362509125324ee277c89a503e04323ce5a70"=>
   "http://mirror.centos.org/centos/7/os/x86_64/images/pxeboot/initrd.img"}]`. The keys are tftp file names/directory and the values are the sources for those files.

Reply:
"Provider is completely unaware where the file will be served". Isn't this the reason where the issue was, as we actually needed the path to be aware of the url?

Reply:
I vote for either keeping @ShimShtein 's proposal, OS-Media needs a substantial overhaul with some mechanism which Katello (or other plugins) could use for overriding, that is not point of this PR.